File: 21. Merge Two Sorted Lists/GenSol.py
'''
21. Merge Two Sorted Lists
https://leetcode.com/problems/merge-two-sorted-lists/description/
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        nn = result = ListNode()
        while list1 and list2:
            if list1.val < list2.val:
                nn.next = list1
                list1, nn = list1.next, list1
            else:
                nn.next = list2
                list2, nn = list2.next, list2
        if list1 or list2:
            nn.next = list1 if list1 else list2

        result = result.next
        
        return result
            

File: 21. Merge Two Sorted Lists/GenSol.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = (list1, list2) => {

    var initialNode = {var:0, next: null}
    var curr = initialNode

    while (list1 && list2) {

        if (list1.val < list2.val) {
            [curr.next, list1] = [list1, list1.next]
        } else {
            [curr.next, list2] = [list2, list2.next]
        }

        curr = curr.next
    }
    curr.next = list1 || list2
    return initialNode.next
};